cornac/models/hypar/hypar.py

```diff
@@ -0,0 +1,965 @@
+
+            # Normalize attention using softmax on edges
+            graph.edata['a'] = self.attn_drop(edge_softmax(graph, e)) # (num_edge, num_heads)
+
+            # If using narre set node representation to original input instead of attention representation
+            if self.aggregator == 'narre':
+                graph.srcdata.update({'el': h_src})
+
+            # Aggregate reviews to nodes.
+            graph.update_all(fn.u_mul_e('el', 'a', 'm'),
+                             fn.sum('m', 'ft'))
+            rst = graph.dstdata['ft']
+
+            # If using activation, apply
+            if self.activation:
+                rst = self.activation(rst)
+
+            # In inference, we may want to get the attention. If so return both review representation and attention.
+            if get_attention:
+                return rst, graph.edata['a']
+            else:
+                return rst
+
+
+class Model(nn.Module):
+    """
+        HypAR model based on DGL and Torch.
+        Parameters
+        ----------
+        g: dgl.DGLGraph
+            Heterogeneous graph with user and item nodes.
+        n_nodes: int
+            Number of nodes
+        aggregator: str
+            Aggregator to use. Can be 'gatv2' and 'narre'.
+        predictor: str
+            Predictor to use. Can be 'narre' and 'dot'.
+        node_dim: int
+            Dimension of node embeddings
+        incidence_dict:
+            Incidence matrix for each relation relation type. I.e., positive and negative AO pairs.
+        num_heads: int
+            Number of heads to use for review aggregation.
+        layer_dropout: list
+            Dropout rate for hypergraph and for review attention layer.
+        attention_dropout: float
+            Dropout rate for attention.
+        preference_module: str
+            Preference module to use. Can be 'lightgcn' and 'mf'.
+        use_cuda: bool
+            Whether we are using cuda.
+        combiner: str
+            Combiner to use. Can be 'add', 'mul', 'bi-interaction', 'concat', 'review-only', 'self', 'self-only'.
+        aos_predictor: str
+            AOS predictor to use. Can be 'non-linear' and 'transr'.
+        non_linear: bool
+            Whether to use non-linear activation function.
+        embedding_type: str
+            Type of embedding to use. Can be 'learned' and 'ao_embeddings'.
+        kwargs: dict
+            Additional arguments, such the learned embeddings.
+        """
+
+    def __init__(self, g, n_nodes, aggregator, predictor, node_dim,
+                 incidence_dict,
+                 num_heads, layer_dropout, attention_dropout, preference_module='lightgcn', use_cuda=True,
+                 combiner='add', aos_predictor='non-linear', non_linear=False, embedding_type='learned',
+                 **kwargs):
+        super().__init__()
+        from cornac.models.lightgcn.lightgcn import Model as lightgcn
```

Comment:
I'm not sure if it's a good practice to import modules from other models. I understand that's code reuse but this also creates a risk of unexpected behaviors if someone changes the implementation of LightGCN later.

Reply:
I suppose the proper solution would be to write a test that ensures lightgcn returns the correct format for the functions. Thus, edits to LightGCN wouldn't affect in/output of functions.

Reply:
I've added the model in commit b8475f3 so no dependencies between HypAR and LightGCN.
I've also removed my edits to LightGCN for this pull request in commit 159968d.